### PR TITLE
Field canonicalization

### PR DESCRIPTION
**Pull request by weinbe58**
- Making waveform IR nodes frozen and therefore hashable. 
    - This involve having to `hack` the casting by using `object.__set_attr__` in the `__post_init__` which isn't idea. Another option for this would be to overload constructor methods that use arbitrary inputs and dispatch those to the default pydantic constructor after casting to scalar values. 
- Merging Scaled locations together if they share the same waveform. 
- Adding `create` function for `PythonFn` node so that Code gen doesn't need to use hacky `object.__set_attr__` to do variable assignment.

### CODE REVIEW

**Comment by Roger-luo** [comment]
I forgot the exact reason, but IIRC there was a reason why we removed the `frozen=True` from the IR objects back in May. Do you remember why was that? I think it was causing trouble for something.

**Comment by weinbe58** [comment]
@Roger-luo Its because Python doesn't have casting. so if you make a `dataclass` with a `Scalar` field there is no way to tell the interpreter how to cast other types to that specific type.

**Comment by Roger-luo** [comment]
I'm confused, how does casting effect mutability here?

**Comment by weinbe58** [comment]
Take the `Constant` node, it takes two arguments `value` and `duration` if I do not overload the constructor and try to do `self.value = cast(value)` I get an error because that field is not mutable.

**Comment by Roger-luo** [comment]
Ah yes, because `dataclass` cannot invoke a type conversion method, that's dumb. I think we should define our own metaclass or custom `field` initializer using `@dataclass_transform` at some point. But yeah I'm convinced with the motivation of this PR now

**Comment by Roger-luo** [comment]
A reference implementation can be found here https://github.com/patrick-kidger/equinox/blob/main/equinox/_module.py#L46

File: src/bloqade/ir/control/field.py [hunk 1]
```diff
@@ -209,9 +218,39 @@ class Field(FieldExpr):
...
 
-    def __hash__(self) -> int:
+    @cached_property
+    def _hash_value(self):
         return hash(frozenset(self.drives.items())) ^ hash(self.__class__)
 
+    def __hash__(self) -> int:
+        return self._hash_value
+
+    def canoncialize(self) -> "Field":
```
Comment (Roger-luo) [inline src/bloqade/ir/control/field.py:228]: put a line of doc here to explain what kind of canonicalization is done here? the spell is wrong btw

**Pull request merged**